packages/vm/src/imports.rs

```diff
@@ -765,6 +771,90 @@ mod test {
         assert_eq!(result, errors::REGION_WRITE_TOO_SMALL);
     }
 
+    #[test]
+    fn do_query_chain_works() {
+        let mut instance = make_instance();
+
+        let request = QueryRequest::Bank(BankQuery::AllBalances {
+            address: HumanAddr::from(INIT_ADDR),
+        });
+        let request_data = cosmwasm_std::to_vec(&request).unwrap();
+        let request_ptr = write_data(&mut instance, &request_data);
+        let response_ptr = create_empty(&mut instance, 1000);
+
+        let ctx = instance.context_mut();
+        leave_default_data(ctx);
+
+        let result = do_query_chain::<S, Q>(ctx, request_ptr, response_ptr);
+        assert_eq!(result, errors::NONE);
+        let response = force_read(ctx, response_ptr);
+
+        let parsed: ApiResult<ApiResult<Binary>, ApiSystemError> =
```

Comment:
That is a long parse chain... maybe we should revisit the query types.
Then again, I think the helper methods on the `Query` trait make this more usable outside of tests

Reply:
What bothers me most with those types is that `ApiResult` is not a `Result` and there is not only an `StdResult` -> `ApiResult` mapping as originally planned, but also some `ApiResult` -> `Result` conversion (I guess to get access to Result methods).

At some point I tried to make `ApiResult<X> = Result<X, ApiError>` but that does not allow trait implementations for `ApiResult`. Maybe we'd be better off by creating an external `fn to_api_result(result: StdResult<X>) -> ApiResult<X>`.

Reply:
Can you open another issue on this one?

I fought with this when updating contracts and then settled on helpers on the Query trait, which make it usable, but only by hiding the ugly interface. If there is a better way to expose this functionality over the wasm-rust boundary, I am very happy to see it.